// zcatalyst-sdk-node/lib/zcql/zcql.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZCQL = void 0;
const validator_1 = require("../utils/validator");
const api_request_1 = require("../utils/api-request");
const constants_1 = __importDefault(require("../utils/constants"));
const error_1 = require("../utils/error");
const { REQ_METHOD, CREDENTIAL_USER, COMPONENT, ACCEPT_HEADER, ZCQL_PARSER } = constants_1.default;
class ZCQL {
    constructor(app) {
        (0, validator_1.isValidApp)(app, true);
        this.requester = new api_request_1.AuthorizedHttpClient(app, this);
    }
    getComponentName() {
        return COMPONENT.zcql;
    }
    executeZCQLQuery(sql) {
        return __awaiter(this, void 0, void 0, function* () {
            yield (0, validator_1.wrapValidatorsWithPromise)(() => {
                (0, validator_1.isNonEmptyString)(sql, 'ZCQL/invalid-argument', true);
            }, error_1.CatalystZCQLError);
            const postData = {
                query: sql
            };
            const request = {
                method: REQ_METHOD.post,
                path: '/query',
                data: postData,
                type: "json" /* JSON */,
                catalyst: true,
                track: true,
                user: CREDENTIAL_USER.user,
                headers: process.env[ZCQL_PARSER] === 'V2' ? { [ACCEPT_HEADER.KEY]: ACCEPT_HEADER.ZCQL } : {}
            };
            const resp = yield this.requester.send(request);
            return resp.data.data;
        });
    }
}
exports.ZCQL = ZCQL;
